src/api/classify_element.rs

```diff
@@ -0,0 +1,349 @@
+}
+
+impl<E: ClassifyElement> ClassifyElementConsumer<E> {
+    fn new(
+        input_stream: ElementStream<E::Packet>,
+        to_providers: Vec<Sender<Option<E::Packet>>>,
+        element: E,
+        await_providers: Vec<Sender<task::Task>>,
+        wake_providers: Vec<Receiver<task::Task>>
+    ) -> Self {
+        ClassifyElementConsumer {
+            input_stream,
+            to_providers,
+            element,
+            await_providers,
+            wake_providers
+        }
+    }
+}
+
+impl<E: ClassifyElement> Drop for ClassifyElementConsumer<E> {
+    fn drop(&mut self) {
+        //TODO: do this with a closure or something, this could be a one-liner
+        for to_provider in self.to_providers.iter() {
+            if let Err(err) = to_provider.try_send(None) {
+                panic!("Consumer: Drop: try_send to provider, fail?: {:?}", err);
+            }
+        }
+
+        for wake_provider in self.wake_providers.iter() {
+            if let Ok(task) = wake_provider.try_recv() {
+                task.notify();
+            }
+        }
+    }
+}
+
+impl<E: ClassifyElement> Future for ClassifyElementConsumer<E> {
+    type Item = ();
+    type Error = ();
+
+    /// Same logic as AsyncElementConsumer, except if any of the channels are full we
+    /// await that channel to clear before processing a new packet. This is somewhat
+    /// inefficient, but seems acceptable for now since we want to yield compute to 
+    /// that producer, as there is a backup.
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        loop {
+            for (port, to_provider) in self.to_providers.iter().enumerate() {
+                if to_provider.is_full() {
+                    let task = task::current();
+                    if let Err(_) = self.await_providers[port].try_send(task) {
+                        task::current().notify();
+                    }
+                    return Ok(Async::NotReady)
+                }
+            }
+            let packet_option: Option<E::Packet> = try_ready!(self.input_stream.poll());
+
+            match packet_option {
+                None => {
+                    return Ok(Async::Ready(()))
+                },
+                Some(packet) => {
+                    let (port, packet) = self.element.classify(packet);
+                    if port >= self.to_providers.len() {
+                        panic!("Tried to access invalid port: {}", port);
+                    }
+                    if let Err(err) = self.to_providers[port].try_send(Some(packet)) {
+                        panic!("Error in to_providers[{}] sender, have nowhere to put packet: {:?}", port, err);
+                    }
+                    if let Ok(task) = self.wake_providers[port].try_recv() {
+                        task.notify();
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// Split Element Provider, exactly the same as AsyncElementProvider, but
```

Comment:
`SplitElementProvider`?

Reply:
Oh needs cleanup, `ClassifyElement` used to be `SplitElement`

Reply:
Can we not modify `AsyncElementProvider` to have the bounds `E: Element + ClassifyElement`?

Reply:
Maybe,  I just didn't want to muck around with it at the time.  It seems possible. Though Classify doesn't do a transformation,  hence the different trait.  Classify implements `fn classify` which outputs essentially a port number with a packet.

Reply:
Maybe @moosingin3space can weigh in?

Reply:
https://github.com/route-rs/route-rs/issues/32